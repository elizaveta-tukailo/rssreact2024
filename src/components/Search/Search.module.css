.search__form {
    display: flex;
    justify-content: flex-end;
}