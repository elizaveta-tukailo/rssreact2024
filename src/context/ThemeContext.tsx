import {
  createContext,
  ReactNode,
  useContext,
  useMemo,
  useState,
  useCallback,
} from 'react';
type Theme = 'light' | 'dark';
interface ThemeContextType {
  theme: Theme;
  toggleTheme: () => void;
}
export const ThemeContext = createContext<ThemeContextType | undefined>(
  undefined
);
export function useTheme() {
  const context = useContext(ThemeContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}
interface ThemeProviderProps {
  children: ReactNode;
}
export function ThemeProvider({ children }: ThemeProviderProps) {
  const [theme, setTheme] = useState<Theme>('light');
  const toggleTheme = useCallback(() => {
    setTheme((prevTheme) => (prevTheme === 'light' ? 'dark' : 'light'));
  }, []);
  const value = useMemo(() => ({ theme, toggleTheme }), [theme, toggleTheme]);
  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
}
