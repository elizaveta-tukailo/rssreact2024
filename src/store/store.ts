import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import { characterApi } from './../services/character';

export const store = configureStore({
  reducer: {
    [characterApi.reducerPath]: characterApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(characterApi.middleware),
});

setupListeners(store.dispatch);
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
