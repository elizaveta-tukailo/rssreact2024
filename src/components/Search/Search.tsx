import { ChangeEvent } from 'react';
import useSearchQuery from '../../hooks/useSearchQuery';
import styles from './search.module.css';
import { useNavigate } from 'react-router-dom';
import { useTheme } from '../../context/ThemeContext';

const Search: React.FC = () => {
  const { theme } = useTheme();
  const [searchQuery, setSearchQuery] = useSearchQuery('searchQuery');
  const navigate = useNavigate();

  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    navigate(`?search=${searchQuery}`);
    localStorage.setItem('searchQuery', searchQuery);
  };

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);
  };

  return (
    <div className={`${styles['searchFormWrap']} ${styles[theme]}`}>
      <form className={styles.searchForm} onSubmit={handleFormSubmit}>
        <input
          className={styles.searchInput}
          type="text"
          name="name"
          value={searchQuery}
          placeholder="Type here"
          onChange={handleChange}
        />
        <button className={styles.searchButton} type="submit">
          Search
        </button>
      </form>
    </div>
  );
};
export default Search;
