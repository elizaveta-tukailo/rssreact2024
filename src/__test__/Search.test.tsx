import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import Search from '../components/Search';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from '../context/ThemeContext';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../store/store';

const mockSetItem = jest.spyOn(Storage.prototype, 'setItem');
const mockGetItem = jest.spyOn(Storage.prototype, 'getItem');

describe('Search', () => {
  it('Show input', () => {
    render(
      <BrowserRouter>
        <ThemeProvider>
          <Search />
        </ThemeProvider>
      </BrowserRouter>
    );
    const input = screen.getByPlaceholderText('Type here');
    expect(input).toBeInTheDocument();
  });
  it('Search component renders correctly', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <ThemeProvider>
              <Search />
            </ThemeProvider>
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Testing Search component', () => {
  afterEach(() => {
    mockSetItem.mockClear();
    mockGetItem.mockClear();
  });
  test('Clicking the Search button saves the entered value to the local storage', () => {
    render(
      <BrowserRouter>
        <ThemeProvider>
          <Search />
        </ThemeProvider>
      </BrowserRouter>
    );
    const input = screen.getByPlaceholderText('Type here') as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'test value' } });
    const button = screen.getByRole('button', { name: /search/i });
    fireEvent.click(button);
    expect(mockSetItem).toHaveBeenCalledWith('searchQuery', 'test value');
  });

  test('Check that the component retrieves the value from the local storage upon mounting', () => {
    mockGetItem.mockImplementation(() => 'stored value');
    render(
      <BrowserRouter>
        <ThemeProvider>
          <Search />
        </ThemeProvider>
      </BrowserRouter>
    );
    const input = screen.getByPlaceholderText('Type here') as HTMLInputElement;
    expect(input.value).toBe('stored value');
  });
});
