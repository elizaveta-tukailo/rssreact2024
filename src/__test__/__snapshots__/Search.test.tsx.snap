// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Card component renders correctly 1`] = `
<div
  className="searchFormWrap light"
>
  <form
    className="searchForm"
    onSubmit={[Function]}
  >
    <input
      className="searchInput"
      name="name"
      onChange={[Function]}
      placeholder="Type here"
      type="text"
      value=""
    />
    <button
      className="searchButton"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;

exports[`Search Search component renders correctly 1`] = `
<div
  className="searchFormWrap light"
>
  <form
    className="searchForm"
    onSubmit={[Function]}
  >
    <input
      className="searchInput"
      name="name"
      onChange={[Function]}
      placeholder="Type here"
      type="text"
      value=""
    />
    <button
      className="searchButton"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
